pkg/asset-manager-utils/contracts/aave/SinglePoolAaveATokenAssetManager.sol

```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import "./ILendingPoolAddressesProvider.sol";
+import "./ILendingPool.sol";
+import "./IAaveIncentivesController.sol";
+
+import "../SinglePoolAssetManager.sol";
+import "@balancer-labs/v2-distributors/contracts/interfaces/IMultiRewards.sol";
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+// solhint-disable no-empty-blocks
+// solhint-disable var-name-mixedcase
+// solhint-disable private-vars-leading-underscore
+contract SinglePoolAaveATokenAssetManager is SinglePoolAssetManager {
+    uint16 private constant REFERRAL_CODE = 0;
+
+    IAaveIncentivesController public immutable aaveIncentives;
+    ILendingPool public immutable lendingPool;
+    IERC20 public immutable aToken;
+    IERC20 public immutable stkAave;
+
+    /// @notice staking contract for pool which owns this asset manager
+    IMultiRewards public distributor;
+
+    constructor(
+        IVault _vault,
+        address _token,
+        address _lendingPool,
+        address _aToken,
+        address _aaveIncentives,
+        address _stkAave
+    ) SinglePoolAssetManager(_vault, bytes32(0), _token) {
+        // TODO: pull these from Aave addresses provider
+        lendingPool = ILendingPool(_lendingPool);
+        aToken = IERC20(_aToken);
+
+        aaveIncentives = IAaveIncentivesController(_aaveIncentives);
+        stkAave = IERC20(_stkAave);
+
+        IERC20(_token).approve(_lendingPool, type(uint256).max);
+    }
+
+    /**
+     * @dev Should be called in same transaction as deployment through a factory contract
+     * @param pId - the id of the pool
+     * @param rewardsContract - the address of the rewards contract (to distribut stkAAVE)
+     */
+    function initialise(bytes32 pId, address rewardsContract) public {
+        require(poolId == bytes32(0), "Already initialised");
```

Comment:
It is very unfortunate that this has to be this way. I don't think we'll even be able to have a single factory contract, since each contract is so large. It'll likely end up being something like the following:

```
contract MegaFactory {
  function createPool(...) {
    assetManager = assetManagerFactory.create(...)
    pool = poolFactory.create(assetMananger, ...);
    assetManager.initialize(pool)
  }
}
```

Even if we end up following that pattern though, I think it may be good to have internal `_initialize` functions for the base contracts (e.g. setting `poolId`) that we call from the derived ones, and remove these arguments (such as `poolId`) from their constructors.